visualizer/exampleGenerationRequests.js

```diff
@@ -0,0 +1,167 @@
+/* eslint-env browser */
+
+'use strict';
+
+// TODO: handle invalid grammar in textbox
+
+(function(root, initModule) {
+  if (typeof exports === 'object') {
+    module.exports = initModule;
+  } else {
+    initModule(root.ohm, root.ohmEditor, root.exampleWorkerManager,
+               root.utils, root.domUtil, root.autosize);
+  }
+})(this, function(ohm, ohmEditor, exampleWorkerManager, utils, domUtil, autosize) {
+
+  var neededExamples = [];
+
+  var focusedElement = null;
+  var focusedRuleName = '';
+
+  // TODO: clicking off of textarea triggers refresh, deleting content
+
+  // logic for arrow to open drawer and show needed examples menu
+  var examplesNeededDrawer = domUtil.$('#examplesNeededDrawer');
+  var examplesNeededIndicator = domUtil.$('#examplesNeededDrawer .indicator');
+  var exampleSplitter = domUtil.$('#exampleSplitter');
+  examplesNeededDrawer.addEventListener('click', function() {
+    var ExampleRequestContainer = domUtil.$('#ExampleRequestContainer');
+    ExampleRequestContainer.classList.toggle('hidden');
+    exampleSplitter.classList.toggle('disabled',
+      ExampleRequestContainer.classList.contains('hidden')
+    );
+  });
+
+  var timeout = null;
+  exampleWorkerManager.addListener('received:neededExamples', function(updatedNeededExamples) {
+    if (timeout) {
+      clearTimeout(timeout);
+    }
+
+    timeout = setTimeout(function() {
+      var neededExampleList = domUtil.$('#exampleRequests ul');
+
+      neededExamples = updatedNeededExamples;
+
+      Array.prototype.slice.call(neededExampleList.children)
+        .forEach(function(childNode) {
+          if (childNode.firstChild !== focusedElement) {
+            neededExampleList.removeChild(childNode);
+          }
+        });
+
+      neededExamples.filter(function(ruleName) {
+        return ruleName !== focusedRuleName;
+      }).forEach(function(ruleName) {
+        var li = domUtil.createElement('li');
+        li.appendChild(makeExampleRequest(ruleName));
+        neededExampleList.appendChild(li);
+      });
+
+      if (neededExamples.length === 0) {
+        examplesNeededIndicator.classList.remove('active');
+      } else {
+        examplesNeededIndicator.classList.add('active');
+      }
+    }, 200);
+  });
+
+  function makeExampleRequest(ruleName) {
+    var request = new ExampleRequest(ohmEditor.grammar, ruleName);
+
+    request.domNode.addEventListener('focusin', function() {
+      focusedElement = request.domNode;
+      focusedRuleName = request.ruleName;
+    });
+    request.domNode.addEventListener('focusout', function(e) {
+      focusedElement = null;
+      focusedRuleName = '';
+
+      if (request.domNode.value.trim() === '' &&
+          !neededExamples.includes(ruleName)) {
+        request.domNode.parentNode.removeChild(request.domNode);
+      }
+    });
+    request.addListener('validSubmit', function(event) {
```

Comment:
In this particular use of the 'ExampleRequest' class, I don't imagine a situation where another listener would want to subscribe to this event. But, If we were to reuse this element at some point, then it would make sense to have multiple subscribers that listen for when a valid item is submitted